### PR TITLE
add VLType and CompoundType, commit complex compound type to file

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [ ] Closes Issue #xxxx
- [x] Tests added
- [x] Changes are documented in `CHANGELOG.rst`
- [x] fix char array issue for compound type

see #228 for transient vs committed types

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
@ZedThree FYI, this is the first serve of adding the missing user types (committed types) to h5netcdf as well as making it compatible with nc-complex.

File: h5netcdf/core.py [hunk 1]
```diff
@@ -724,6 +748,21 @@ def _create_child_variable(
...
+        # compound type handling
+        if np.dtype(dtype).kind == "c":
+            if not isinstance(dtype, CompoundType) and self._root._auto_complex:
+                # todo check compound types for
+                #  existing complex types which may be used
+                itemsize = np.dtype(dtype).itemsize
+                if itemsize == 8:
+                    width = "FLOAT"
+                elif itemsize == 16:
+                    width = "DOUBLE"
```
Comment (ZedThree) [inline h5netcdf/core.py:760]: I don't know how common other type sizes are, but it's probably wise to explicitly error here

**Comment by ZedThree** [comment]
LGTM! I think your xarray tests are failing because, somewhat ironically, they actually work and you expect them to fail! :smile: 

Just to note that nc-complex supports a wider variety of conventions for complex numbers, but I think it's absolutely fine for h5netcdf to only support the h5py/transient compound type and the explicit `_PFNC_*_COMPLEX` named type. h5netcdf will be able to read them all perfectly ok, just not auto-detect them as being complex numbers.

**Comment by kmuehlbauer** [comment]
Thanks for taking a look here, much appreciated.

> LGTM! I think your xarray tests are failing because, somewhat ironically, they actually work and you expect them to fail! 😄

Yes, they are actually from xarray's own test-suite, which expect h5netcdf to error here :grinning: 

> Just to note that nc-complex supports a wider variety of conventions for complex numbers, but I think it's absolutely fine for h5netcdf to only support the h5py/transient compound type and the explicit `_PFNC_*_COMPLEX` named type. h5netcdf will be able to read them all perfectly ok, just not auto-detect them as being complex numbers.

Sure, we can think about aligning h5netcdf's legacyapi to what netcdf4-python supports.

**Comment by ZedThree** [comment]
> Sure, we can think about aligning h5netcdf's legacyapi to what netcdf4-python supports.

This would be neat, but much less important than your work here to ensure that currently released (and older) versions of netcdf will be able to read files created by h5netcdf. This is a great improvement for compatibility, so thank you!

**Comment by kmuehlbauer** [comment]
@shoyer, I'd like to merge this in the coming days. If you have time it would be great if you could take a look here. Thanks!

With the work of @ZedThree over in https://github.com/PlasmaFAIR/nc-complex and this PR here, we can enable complex numbers for xarray netcdf4/h5netcdf backends.

File: h5netcdf/core.py (hunk 1 above)
Comment (shoyer) [reply h5netcdf/core.py:760]: `width = {8: "FLOAT", 16: "DOUBLE"}[itemsize]` still feels like a safer way to write this tp me :)

File: h5netcdf/core.py [hunk 2]
```diff
@@ -1127,6 +1211,7 @@ def __init__(self, path, mode="r", invalid_netcdf=False, phony_dims=None, **kwar
         track_order = kwargs.pop("track_order", track_order_default)
 
         self.decode_vlen_strings = kwargs.pop("decode_vlen_strings", None)
+        self._auto_complex = kwargs.pop("auto_complex", None)
```
Comment (shoyer) [inline h5netcdf/core.py:1214]: Why not use real keyword arguments here (which could be documented in the docstring) instead of putting them in `**kwargs`?

File: h5netcdf/core.py [hunk 3]
```diff
@@ -1192,9 +1277,17 @@ def __init__(self, path, mode="r", invalid_netcdf=False, phony_dims=None, **kwar
...
+            # this creates files which are readable by netCDF4 with nc-complex
+            # as well as h5netcdf
+            if self._auto_complex is None:
+                self._auto_complex = True
         else:
             if self.decode_vlen_strings is None:
                 self.decode_vlen_strings = False
+            # default _auto_complex to False for new API
+            if self._auto_complex is None:
+                self._auto_complex = False
```
Comment (shoyer) [inline h5netcdf/core.py:1290]: This mismatch between the legacy and new APIs is a little strange and could surprise users. A more conservative choice (which could be changed later) is to defautl to `auto_complex=False` and require users to explicitly opt-in to the new behavior.

Are arrays written with `auto_complex=True` automatically read as complex valued with h5py or old versions of h5netcdf? If not, I would stick with `auto_complex=False` as a default in both cases until a concensus emerges that this is the right way to handle complex values in netCDF files.

File: h5netcdf/tests/test_h5netcdf.py [hunk 4]
```diff
@@ -2404,6 +2404,181 @@ def test_enum_type_creation(tmp_local_or_remote_netcdf, netcdf_write_module):
...
+    complex_array = np.array([0 + 0j, 1 + 0j, 0 + 1j, 1 + 1j, 0.25 + 0.75j])
+    kwargs = dict(auto_complex=True)
+
+    with netcdf_write_module.Dataset(tmp_local_netcdf, "w", **kwargs) as ds:
+        ds.createDimension("x", size=len(complex_array))
+        if dtype == "c4":
+            try:
+                var = ds.createVariable("data", dtype, ("x",))
+            except TypeError as e:
+                assert e.args[0] == "data type 'c4' not understood"
```
Comment (shoyer) [inline h5netcdf/tests/test_h5netcdf.py:2564]: pytest.raises is the standard way to write this: https://seaborn.pydata.org/tutorial/color_palettes.html#perceptually-uniform-palettes

File: h5netcdf/tests/test_h5netcdf.py [hunk 5]
```diff
@@ -2404,6 +2404,181 @@ def test_enum_type_creation(tmp_local_or_remote_netcdf, netcdf_write_module):
...
+                if netcdf_write_module.__name__ == "netCDF4":
+                    assert "Illegal primitive data type" in e.args[0]
+                else:
+                    assert (
+                        e.args[0]
+                        == "Currently only 'complex64' and 'complex128' dtypes are allowed."
+                    )
+        else:
+            var = ds.createVariable("data", dtype, ("x",))
+            var[:] = complex_array
```
Comment (shoyer) [inline h5netcdf/tests/test_h5netcdf.py:2579]: Generally it's cleaner to write separate unit tests rather than a parameterized test with branched logic.

File: h5netcdf/core.py (hunk 2 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py:1214]: I was hesitant to do this. But it will surely be more visible, thanks!

File: h5netcdf/core.py (hunk 3 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py:1290]: The only difference for h5py/h5netcdf is the additional committed (named) type within the file on write. Old versions can read such files, relying on the type retrieved from the dataset. New versions of h5netcdf also gather the type in `.cmptypes`.

But going conservative first is a good idea anyway. netcdf4-python does the same, @ZedThree, right?

Thanks Stefan for the review, much appreciated.  😊

**Comment by kmuehlbauer** [comment]
I could not get around to check this in more detail in this gist: https://gist.github.com/kmuehlbauer/1f6f8a61ad5b0a7693ae08154060b790

It looks like we could get away by just committing the complex compound type to file as nectdf-c/netcdf4-python does instead of relying on the transient hdf5 type. This could also be considered as bug.

That means we could significantly simplify this PR. We would not have to use the `auto_complex` keyword, since we would write perfectly sane NetCDF4 files. All these files are readable with former netcdf-c/netcdf4-python versions as well as h5netcdf versions. This would be fully backwards compatible.

File: h5netcdf/core.py (hunk 2 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Not needed with latest approach.

File: h5netcdf/core.py (hunk 3 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: I've checked again, see https://github.com/h5netcdf/h5netcdf/pull/227#issuecomment-1902243420

File: h5netcdf/core.py [hunk 6]
```diff
@@ -724,6 +748,23 @@ def _create_child_variable(
...
+            if not isinstance(dtype, CompoundType):
+                # todo check compound types for
+                #  existing complex types which may be used
+                itemsize = np.dtype(dtype).itemsize
+                try:
+                    width = {8: "FLOAT", 16: "DOUBLE"}[itemsize]
+                except KeyError:
+                    raise TypeError(
+                        "Currently only 'complex64' and 'complex128' dtypes are allowed."
+                    )
```
Comment (shoyer) [inline h5netcdf/core.py:762]: This results in a slightly better traceback ("direct cause of" rather than "during the handling of")
```suggestion
                except KeyError as e:
                    raise TypeError(
                        "Currently only 'complex64' and 'complex128' dtypes are allowed."
                    ) from e
```

File: h5netcdf/core.py [hunk 7]
```diff
@@ -842,7 +882,7 @@ def _create_child_variable(
         # when a variable is first written to, after variable creation.
         # Here we just attach it to every variable on creation.
         # Todo: get this consistent with netcdf-c/netcdf4-python
-        variable._ensure_dim_id()
+        # variable._ensure_dim_id()
```
Comment (shoyer) [inline h5netcdf/core.py:885]: This also looks unrelated.

File: h5netcdf/core.py [hunk 8]
```diff
@@ -246,7 +259,7 @@ def _attach_coords(self):
             [self._parent._all_dimensions[d]._dimid for d in dims],
             "int32",
         )
-        if len(coord_ids) > 1:
+        if len(coord_ids) >= 1:
```
Comment (shoyer) [inline h5netcdf/core.py:262]: This looks unrelated

Comment (kmuehlbauer) [reply h5netcdf/core.py]: Thanks!

File: h5netcdf/core.py (hunk 7 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Thanks!

**Comment by kmuehlbauer** [comment]
@shoyer Thanks for the immediate review. I've fixed the mentioned issues and applied your change for exception handling.

Sure, xarray's h5netcdf tests (https://github.com/h5netcdf/h5netcdf/actions/runs/7597039010/job/20691426586?pr=227) are breaking, because of the complex compatibility tests over there. This will immediately lead to breaking nightly tests over at xarray, if this PR is merged. I'm ready to take action over at xarray.

**Comment by kmuehlbauer** [comment]
Unfortunately there are still some inconsistencies in the correct retrieval of the committed type. I'll push more changes later.

File: h5netcdf/core.py [hunk 9]
```diff
@@ -724,6 +748,26 @@ def _create_child_variable(
...
+                        "Currently only 'complex64' and 'complex128' dtypes are allowed."
+                    ) from e
+                dname = f"_PFNC_{width}_COMPLEX_TYPE"
+                dtype = self._all_cmptypes.get(
+                    dname, self.create_cmptype(np.dtype(dtype), dname)
+                )
+            # use committed type
+            # that will actually reference the named type in the dataset
+            dtype = dtype._h5ds
+
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:770]: This actually references the committed type within the dataset. I've used `h5ls` to check this. This tool was recommended at hdf5forum. For some reason I can't retrieve the `name` of the committed type from the dataset datatype. I've struggled to retrieve this from the plain c code over at hdf5 but that did not work out. 

There is one solution to check the dataset type with any committed type for equality, but this seems quite clunky. I'll implement that solution to move forward here.

Comment (kmuehlbauer) [reply h5netcdf/core.py:770]: Just before giving up, I've found another way of retrieving the name of the linked type directly from the dataset (here `datac2`):

```python
ds._root._h5py.h5i.get_name(datac2._h5ds.id.get_type()).decode()
```

Getting back to this the next days. @shoyer Sorry, but this is more involved than I anticipated.

File: h5netcdf/core.py [hunk 10]
```diff
@@ -258,6 +271,28 @@ def _ensure_dim_id(self):
             if "_Netcdf4Dimid" in dim.attrs:
                 self._h5ds.attrs["_Netcdf4Dimid"] = dim.attrs["_Netcdf4Dimid"]
 
+    def _add_fillvalue(self, fillvalue, dtype):
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:274]: Refactored out from create_child_variable

File: h5netcdf/core.py [hunk 11]
```diff
@@ -316,6 +351,27 @@ def __len__(self):
     @property
     def datatype(self):
         """Return numpy dtype or user defined type."""
+        if self._root._h5py.__name__ == "h5py":
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:354]: This is the current procedure to extract the committed type as h5netcdf type. For h5pyd and transient types special handling is needed.

File: h5netcdf/core.py [hunk 12]
```diff
@@ -524,10 +582,93 @@ def _unlabeled_dimension_mix(h5py_dataset):
     return status
 
 
+def _check_h5type(self, dtype, np_dtype):
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:585]: Refactored out from create_child_variable.

File: h5netcdf/core.py [hunk 13]
```diff
@@ -524,10 +582,93 @@ def _unlabeled_dimension_mix(h5py_dataset):
...
+        # create and commit type in current group
+        if dname not in self._all_cmptypes:
+            self.create_cmptype(np_dtype, dname)
+        # get committed type from file
+        return self._all_cmptypes[dname]._h5ds
+    else:
+        return dtype
+
+
+def _check_fillvalue(self, fillvalue, dtype, np_dtype):
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:638]: Refactored out from create_child_variable

File: h5netcdf/core.py [hunk 14]
```diff
@@ -790,32 +956,21 @@ def _create_child_variable(
...
+        # use numpy dtype for h5pyd
+        if self._root._h5py.__name__ == "h5pyd":
+            h5type = np_dtype
 
         # create hdf5 variable
         self._h5group.create_dataset(
             h5name,
             shape,
-            dtype=dtype,
+            dtype=h5type,
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:970]: Using the h5type (for user types this would be the committed type) this will link/share the type in the dataset.

File: h5netcdf/core.py [hunk 15]
```diff
@@ -258,6 +271,28 @@ def _ensure_dim_id(self):
...
+                and string_info.length is not None
+                and string_info.length > 1
+            ) or enum_info:
+                value = fillvalue
+            else:
+                value = self.dtype.type(fillvalue)
+
+        # need to use create-function in order
+        # to provide correct committed/named type
+        self.attrs._h5attrs.create("_FillValue", value, dtype=dtype)
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:294]: Specifying the dtype here (for user types this would be the committed type) will link/share the type in the attribute.

**Review by kmuehlbauer — COMMENTED** [review]
I've added some comments for better understanding.

File: h5netcdf/core.py (hunk 15 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py:294]: Note to self: Might need to special case `_FillValue` in `attrs.py` `__setitem__` to write it with dtype (committed type) from variable in any case (user might overwrite `_FillValue` attribute)

**Comment by kmuehlbauer** [comment]
This seems to work now as intended. I'll split out the changes regarding EnumType and shared user type into a separate PR's. This will finally wrap up VLType and CompoundType.

File: h5netcdf/core.py [hunk 16]
```diff
@@ -578,14 +592,31 @@ def _check_dtype(self, dtype):
         # check if committed type can be accessed in current group hierarchy
         dname = dtype.name.split("/")[-1]
         if (
-            (user_type := self._all_enumtypes.get(dname)) is None
-        ) or user_type._h5ds.name != h5type.name:
+            (user_type := self._all_usertypes.get(dname)) is None
+        ) or self._root._h5file[user_type].name != h5type.name:
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:596]: This looks very ugly. I'll think about another approach to retrieve this.

File: h5netcdf/core.py [hunk 17]
```diff
@@ -691,15 +724,25 @@ def __init__(self, parent, name):
 
         self._dimensions = Dimensions(self)
         self._enumtypes = _LazyObjectLookup(self, self._enumtype_cls)
+        self._vltypes = _LazyObjectLookup(self, self._vltype_cls)
+        self._cmptypes = _LazyObjectLookup(self, self._cmptype_cls)
+        self._usertypes = dict()
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:729]: Unfortunately we can't utilize _LazyObjectLookup in a neat manner for a joined usertype-dict. Using a simple dict for now, just adding name/h5name key/value-pairs.

File: h5netcdf/core.py [hunk 18]
```diff
@@ -713,11 +756,14 @@ def __init__(self, parent, name):
...
-            ) and self._root._h5py.check_enum_dtype(v.dtype):
-                self._enumtypes.add(k)
+            elif isinstance(v, self._root._h5py.Datatype):
+                if self._root._h5py.check_enum_dtype(v.dtype):
+                    self._enumtypes.add(k)
+                elif self._root._h5py.check_vlen_dtype(v.dtype):
+                    self._vltypes.add(k)
+                elif v.dtype.names is not None or "complex" in v.dtype.name:
+                    self._cmptypes.add(k)
+                self._usertypes[k] = v.name
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:766]: If someone has a nice idea how to remove that chain of checks...

File: h5netcdf/core.py [hunk 19]
```diff
@@ -1175,9 +1233,44 @@ def create_enumtype(self, datatype, datatype_name, enum_dict):
         self._h5group[datatype_name] = et
         # create enumtype class instance
         enumtype = self._enumtype_cls(self, datatype_name)
+        # enumtype = self._usertype_cls(self, datatype_name)
         self._enumtypes[datatype_name] = enumtype
+        self._usertypes[datatype_name] = enumtype._h5ds.name
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:1238]: Adding the type to lazy-dict and the h5name to usertype-dict. Same thing below.

File: h5netcdf/core.py [hunk 20]
```diff
@@ -578,14 +592,31 @@ def _check_dtype(self, dtype):
...
+            ) from e
+        dname = f"_PFNC_{width}_COMPLEX_TYPE"
+        # todo check compound type for existing complex types
+        #  which may be used her
+        # if dname is not available in current group-path
+        # create and commit type in current group
+        if dname not in self._all_cmptypes:
+            dtype = self.create_cmptype(dtype, dname).dtype
+        # get committed type from file
+        # return self._all_cmptypes[dname]._h5ds
```
Comment (kmuehlbauer) [inline h5netcdf/core.py:619]: This is the complex compound creation part. This is really backwards compatible, as h5netcdf always reads these complex compound types as complex numbers. netcdf-c/netcdf4-python can read those compounds if, like we do now, the complex compound type is committed to the file.

**Review by kmuehlbauer — COMMENTED** [review]
@shoyer @ZedThree

After #229 and #230 (some refactoring to keep good review experience), I've updated this PR with the undisputed changes to get VL and Compound types working for h5netcdf. Plays well together with nc-complex AFAICT.

I'll do some more clean-up the next days, but you might already have a look here.

File: h5netcdf/core.py (hunk 16 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Fixed!

File: h5netcdf/core.py (hunk 17 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Fixed!

File: h5netcdf/core.py (hunk 18 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Fixed!

File: h5netcdf/core.py (hunk 19 above)
Comment (kmuehlbauer) [reply h5netcdf/core.py]: Fixed! Not needed anymore!

**Comment by kmuehlbauer** [comment]
I think this is ready for final review. I really like to get this in sometime next week, and get a release out.

**Comment by kmuehlbauer** [comment]
So, this is unfortunately not finished yet. There are issues with compound type containing strings. This is done by array of chars under the hood, which needs some special casing to be transparent in h5netcdf. See: https://github.com/Unidata/netcdf4-python/issues/773

I'll set this PR to draft until this is figured out.

**Comment by kmuehlbauer** [comment]
@ZedThree @shoyer Finally I got around to finish this up. Please have a look if you are around.

I'm having another self-review early next week and will merge after that.

**Comment by ZedThree** [comment]
LGTM! Thanks @kmuehlbauer !
